### PR TITLE
Redis sessions

### PR DESCRIPTION
**Pull request by zacharygrafton**
Hi David,

I added Redis session support. Not sure if you were interested, but I thought other might be.

Thanks,
Zack

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -11,6 +11,8 @@ SET(ONION_USE_JPEG true CACHE BOOL "Adds support for simple image handler")
 SET(ONION_USE_XML2 true CACHE BOOL "Adds support for XML2 lib, which is needed for WebDAV handler")
 SET(ONION_USE_SYSTEMD true CACHE BOOL "Adds simple support for systemd")
 SET(ONION_USE_SQLITE3 true CACHE BOOL "Adds support for sqlite3 sessions")
+SET(ONION_USE_REDIS true CACHE BOOL "Adds support for redis sessions")
+SET(ONION_USE_REDIS true CACHE BOOL "Adds support for redis sessions")
```
Comment (davidmoreno) [inline CMakeLists.txt]: This line looks duplicated, isn't it?

**Comment by davidmoreno** [comment]
Wow! Thanks a lot it can be indeed very useful.

Apart from the one line coment I'm worried that the session_id is passed as is to the library, and inside I don't know if there is any validation, but if there is not, it may well cause some 'redis injection' similar in damage to SQL injections. Maybe a quick only alnum allowed characters in the session id can be added to fix this problem.

But I don't know if the library takes care of it. If it does, a single comment would be nice.

Regards, 
David

**Comment by zacharygrafton** [comment]
David,

I removed the duplicate line, however, I'm unsure of where to place the comment about the way that hiredis handles the string interpolation. From their documentation, and looking at the code, the redisCommand function does quite a bit or work on parameters passed to it while translating the input to the wire protocol. I imagine it would be pretty challenging to exploit the code in an effective manner.

Zack

**Comment by davidmoreno** [comment]
Hi,

I think better safe than sorry, so we should do something like:

char *p=session_id;
while(*p){
  if (!isalnum(*p))
     return NULL;
  p++;
}

(Wrote from my mobile, this is just the main idea)

At every possible use of session id.

This way for sure it is something session id like. If not as I said user
could add a ;, \n space or whatever character that redis use to separate
commands and deliver two commands.

I think it's worth doing it.

Regards,
David

**Comment by zacharygrafton** [comment]
David,

Redis uses an interesting protocol where all strings are passed in as binary safe strings. I missed that bit when using specifiers with hiredis library. So, I changed all the specifiers to use %b which means the value should be passed as a binary safe string. There is absolutely no risk of using the session_id in a harmful manner. Also, it isn't possible to specify more than one command at a time in Redis. Inserting something like a ";" or a "\n" won't change anything now as it will all be considered binary data by Redis.

Zack

**Comment by davidmoreno** [comment]
Great then. I will merge in a moment.

Thanks!

**Pull request merged**